public/app/features/auth-config/components/ServerDiscoveryField.tsx

```diff
@@ -0,0 +1,73 @@
+import { useState } from 'react';
+import { UseFormSetValue } from 'react-hook-form/dist/types/form';
+
+import { AppEvents } from '@grafana/data';
+import { getAppEvents, getBackendSrv } from '@grafana/runtime';
+import { Button } from '@grafana/ui';
+
+import { Trans } from '../../../core/internationalization';
+import { ServerDiscoveryFormData, SSOProviderDTO } from '../types';
+
+import { ServerDiscoveryModal } from './ServerDiscoveryModal';
+
+interface Props {
+  setValue: UseFormSetValue<SSOProviderDTO>;
+}
+
+export const ServerDiscoveryField = ({ setValue }: Props) => {
+  const appEvents = getAppEvents();
+  const [isModalOpen, setModalVisibility] = useState(false);
+  const onClose = () => setModalVisibility(false);
+  const onSuccess = async (data: ServerDiscoveryFormData) => {
+    try {
+      const wellKnownSuffix = '/.well-known/openid-configuration';
+      const url = new URL(data.url);
+      if (!url.pathname.includes(wellKnownSuffix)) {
+        data.url = url.origin + wellKnownSuffix;
+      }
+
+      const res = await getBackendSrv().get(data.url);
+
+      if (!res['token_endpoint'] || !res['authorization_endpoint']) {
+        appEvents.publish({
+          type: AppEvents.alertWarning.name,
+          payload: ['The URL provided is not a valid .well-known/openid-configuration endpoint'],
+        });
+        return;
+      }
+
+      setValue('tokenUrl', res['token_endpoint']);
+      setValue('authUrl', res['authorization_endpoint']);
+      if (res['userinfo_endpoint']) {
+        setValue('apiUrl', res['userinfo_endpoint']);
+      }
+
+      appEvents.publish({
+        type: AppEvents.alertSuccess.name,
+        payload: ['OpenID Connect Discovery URL has been successfully fetched.'],
+      });
+    } catch (error) {
+      appEvents.publish({
+        type: AppEvents.alertWarning.name,
+        payload: ['Failed to fetch URL or invalid content'],
+      });
+    } finally {
+      onClose();
+    }
+  };
+  return (
+    <>
+      <Button
+        type="button"
+        fill="text"
+        variant="secondary"
+        onClick={() => {
+          setModalVisibility(true);
+        }}
+      >
+        <Trans i18nKey={'oauth.form.server-discovery-action-button'}>Enter OpenID Connect Discovery URL</Trans>
```

Comment:
Should this text be updated after a URL is set? E.g., `Update OpenID Connect Discovery URL` to show that a URL is already set.

Reply:
I'm not sure we should (or even can) do this because the value for the discovery field is not saved in the database. We cannot tell whether a value has been entered before when a user opens the generic oauth form.